### PR TITLE
daemon/config: New(): implement platform-specific defaults

### PR DESCRIPTION
**Pull request by thaJeztah**
- extracting more bits from https://github.com/moby/moby/pull/43703

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/config/options.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package config // import "github.com/docker/docker/daemon/config"
+
+// WithDefaults initializes platform-agnostic default values for the Config.
+func WithDefaults(cfg *Config) {
```
Comment (tianon) [inline daemon/config/options.go:4]: If this is only used by `New()` (and unconditionally at that), does it need to be exported?

**Review by tianon — COMMENTED** [review]
This is an extremely rare case of my not understanding the split between your commits :innocent:

They seem very tightly coupled IMO -- the first isn't terribly useful/transformative without the second. :see_no_evil:

Overall the change seems sane to me, although it seems a little bit weird to have the defaults split into two different places (depending on how dynamic the default value is, as I understand it?) -- won't that make it a little harder to determine which one is "canonical" in the future (thinking of future contributors who are unfamiliar with this change, or myself in 15-20 minutes when I've forgotten about it)? :grimacing:

File: daemon/config/options.go (hunk 1 above)
Comment (thaJeztah) [reply daemon/config/options.go:4]: Ah, yes, I guess I can un-export it

**Comment by thaJeztah** [comment]
>  it seems a little bit weird to have the defaults split into two different places (depending on how dynamic the default value is, as I understand it?) -- won't that make it a little harder to determine which one is "canonical" in the future (thinking of future contributors who are unfamiliar with this change, or myself in 15-20 minutes when I've forgotten about it)? 


Yes, I've been struggling a bit with that part. Initially wanted to make those bits a functional argument for `config.New()`; I think there was a bit of a chicken and egg situation there (need to check again); besides some of the "dynamic" options being rather heavy-weight, which made me doubt if they all should be part of `config.New()`.

Let me try a bit what it would look like (if possible) to move those bits to the `config` package as well

**Comment by thaJeztah** [comment]
@tianon I pushed a (temporary) commit that updates this PR to;

- remove the functional arguments
- moves all (platform) defaults to the `daemon/config` package
- makes the (platform) defaults internal to `config.New()`
- adds an error-return to `config.New()` (for situations where defaults can produce an error)

I agree it looks better / less complicated; I'll squash the commits if you like the last iteration 👍

**Review by tianon — APPROVED** [review]
Nice, I like it!  All the options grouped together very cleanly. :+1:

**Comment by thaJeztah** [comment]
thx! I'll do a squash tomorrow 👍

**Comment by thaJeztah** [comment]
Squishy-squashy done; let me move this one out of draft again 👍

**Comment by thaJeztah** [comment]
@corhere @cpuguy83 @AkihiroSuda PTAL

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**